core/src/tasks/settle_match_internal.rs

```diff
@@ -0,0 +1,272 @@
+    pub fn new(
+        execution_price: FixedPoint,
+        order1: OrderIdentifier,
+        order2: OrderIdentifier,
+        order1_proof: OrderValidityProofBundle,
+        order1_witness: OrderValidityWitnessBundle,
+        order2_proof: OrderValidityProofBundle,
+        order2_witness: OrderValidityWitnessBundle,
+        match_result: MatchResult,
+        starknet_client: StarknetClient,
+        network_sender: TokioSender<GossipOutbound>,
+        global_state: RelayerState,
+        proof_manager_work_queue: CrossbeamSender<ProofManagerJob>,
+    ) -> Self {
+        Self {
+            execution_price,
+            order_id1: order1,
+            order_id2: order2,
+            order1_proof,
+            order1_validity_witness: order1_witness,
+            order2_proof,
+            order2_validity_witness: order2_witness,
+            match_result,
+            valid_match_mpc: None,
+            starknet_client,
+            network_sender,
+            global_state,
+            proof_manager_work_queue,
+            task_state: SettleMatchInternalTaskState::Pending,
+        }
+    }
+
+    /// Prove `VALID MATCH MPC` on the order pair
+    async fn prove_match_mpc(&mut self) -> Result<(), SettleMatchInternalTaskError> {
+        // Build a witness
+        let commitment_witness1 = self.order1_validity_witness.copy_commitment_witness();
+        let commitment_witness2 = self.order2_validity_witness.copy_commitment_witness();
+
+        let witness = ValidMatchMpcWitness {
+            order1: commitment_witness1.order.clone(),
+            balance1: commitment_witness1.balance_send.clone(),
+            amount1: self.match_result.base_amount.into(),
+            price1: self.execution_price,
+            order2: commitment_witness2.order.clone(),
+            balance2: commitment_witness2.balance_send.clone(),
+            amount2: self.match_result.base_amount.into(),
+            price2: self.execution_price,
+            match_res: self.match_result.clone().to_linkable(),
+        };
+
+        // Enqueue a job with the proof generation module
+        let (response_sender, response_receiver) = oneshot::channel();
+        self.proof_manager_work_queue
+            .send(ProofManagerJob {
+                type_: ProofJob::ValidMatchMpcSingleprover { witness },
+                response_channel: response_sender,
+            })
+            .map_err(|_| {
+                SettleMatchInternalTaskError::EnqueuingJob(ERR_ENQUEUING_JOB.to_string())
+            })?;
+
+        // Await the proof from the proof manager
+        let proof = response_receiver.await.map_err(|_| {
+            SettleMatchInternalTaskError::EnqueuingJob(ERR_AWAITING_PROOF.to_string())
+        })?;
+        self.valid_match_mpc = Some(proof.into());
+
+        Ok(())
+    }
+
+    /// Prove `VALID SETTLE` on the order pair
+    async fn prove_settle(&self) -> Result<(), SettleMatchInternalTaskError> {
+        todo!()
```

Comment:
thought for the future: after we successfully create a `VALID SETTLE` proof, we should shoot down in-flight MPCs for both of the matched orders.

or, does this happen automatically when the wallet/order states are updated?

Reply:
Yeah this is a good point, it does happen automatically when the nullifier is spent on-chain, but we can do this optimistically since it's an internal match. I'll add a TODO